utils/src/server/helpers.rs

```diff
@@ -82,67 +81,108 @@ impl<'a> io::Write for &'a mut BoundedWriter {
 	}
 }
 
-/// Helper for sending JSON-RPC responses to the client
-pub fn send_response(id: Id, tx: &MethodSink, result: impl Serialize, max_response_size: u32) {
-	let mut writer = BoundedWriter::new(max_response_size as usize);
+/// Sink that is used to send back the result to the server for a specific method.
+#[derive(Clone, Debug)]
+pub struct MethodSink {
+	/// Channel sender
+	tx: mpsc::UnboundedSender<String>,
+	/// Max response size in bytes for a executed call.
+	max_response_size: u32,
+}
 
-	let json = match serde_json::to_writer(&mut writer, &Response { jsonrpc: TwoPointZero, id: id.clone(), result }) {
-		Ok(_) => {
-			// Safety - serde_json does not emit invalid UTF-8.
-			unsafe { String::from_utf8_unchecked(writer.into_bytes()) }
-		}
-		Err(err) => {
-			tracing::error!("Error serializing response: {:?}", err);
-
-			if err.is_io() {
-				let data = to_json_raw_value(&format!("Exceeded max limit {}", max_response_size)).ok();
-				let err = ErrorObject {
-					code: ErrorCode::ServerError(OVERSIZED_RESPONSE_CODE),
-					message: OVERSIZED_RESPONSE_MSG.into(),
-					data: data.as_deref(),
-				};
-				return send_error(id, tx, err);
-			} else {
-				return send_error(id, tx, ErrorCode::InternalError.into());
+impl MethodSink {
+	/// Create a new `MethodSink` with unlimited response size
+	pub fn new(tx: mpsc::UnboundedSender<String>) -> Self {
+		MethodSink { tx, max_response_size: u32::MAX }
+	}
+
+	/// Create a new `MethodSink` with a limited response size
+	pub fn new_with_limit(tx: mpsc::UnboundedSender<String>, max_response_size: u32) -> Self {
+		MethodSink { tx, max_response_size }
+	}
+
+	/// Send a JSON-RPC response to the client. If the serialization of `result` exceeds `max_response_size`,
+	/// an error will be sent instead.
+	pub fn send_response(&self, id: Id, result: impl Serialize) -> bool {
+		let mut writer = BoundedWriter::new(self.max_response_size as usize);
+
+		let json = match serde_json::to_writer(&mut writer, &Response { jsonrpc: TwoPointZero, id: id.clone(), result })
+		{
+			Ok(_) => {
+				// Safety - serde_json does not emit invalid UTF-8.
+				unsafe { String::from_utf8_unchecked(writer.into_bytes()) }
 			}
-		}
-	};
+			Err(err) => {
+				tracing::error!("Error serializing response: {:?}", err);
+
+				if err.is_io() {
+					let data = to_json_raw_value(&format!("Exceeded max limit {}", self.max_response_size)).ok();
+					let err = ErrorObject {
+						code: ErrorCode::ServerError(OVERSIZED_RESPONSE_CODE),
+						message: OVERSIZED_RESPONSE_MSG.into(),
+						data: data.as_deref(),
+					};
+					return self.send_error(id, err);
+				} else {
+					return self.send_error(id, ErrorCode::InternalError.into());
```

Comment:
I assume that the max size error is an error based on user input being too large?

I wonder whether the `tracing::error!("Error serializing response: {:?}", err);` line should instead be above this InternalError bit, so we only get output for the unexpected errors?

Reply:
It means that `output` from an executed call exceeded the max limit but the "request" itself was below the limit.
Thus, it depends what you mean with the user but the `registered callback` created a response bigger than the max limit.

Yes, or downgrade the error above to warn or something but the `InternalError` "should" be infallible/bug IIRC.